### PR TITLE
Add workflow that compares to the upstream observatory code and trigg…

### PR DESCRIPTION
**Pull request by moeyensj**
…ers a package publication on changes

### CODE REVIEW

**Comment by moeyensj** [comment]
I've tested this on my local system and can confirm that the build, test, and publish workflow is correctly triggered on a change with this new workflow. The real test will be to see if this works in production.

File: .github/workflows/compare_upstream.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    steps:
+    - name: Checkout Repository
+      uses: actions/checkout@v3
+    - name: Set up Python 3.11
+      uses: actions/setup-python@v3
+      with:
+        python-version: "3.11"
+    - name: Upgrade pip and install mpc_obscodes
+      run: python -m pip install --upgrade build requests mpc-obscodes
```
Comment (akoumjian) [inline .github/workflows/compare_upstream.yml:21]: We may encounter problems both checking out the package using `actions/checkout` and then pip installing the latest version from PyPI. 

Might I recommend using pip download and doing something like this instead:

```
# Download the contents of the PyPI package to the `remote` folder
pip download --no-binary :all: -d remote mpc-obscodes

# Extract the file we want to md5 compare and store it in `existing/obscodes_extended.json`
# You may need to adjust this command if running on macOS
tar -zxf remote/* --wildcards --no-anchored '*obscodes_extended.json' -O > existing/obscodes_extended.json

# Fetch the new files we care about and stick them in `new`
python mpc_obscodes/fetch.py --compressed_file new/obscodes_extended.json.gz --decompressed_file new/obscodes_extended.json --md5_file new/obscodes_extended_new.md5

# Compare the md5s of the two files and indicate upstream change in the output
if [ "$(md5sum existing/obscodes_extended.json new/obscodes_extended.json | awk '{print $1}' | uniq -c | grep -c "1 ")" -eq 1 ]; then export GITHUB_OUTPUT+="upstream_change=true "; fi
```

**Review by akoumjian — APPROVED** [review]
Since we don't package the `tar` or the `md5` files, I might recommend removing those as options in the script and simply have a single file out as an option.

I left some feedback but it's non-blocking.

File: .github/workflows/compare_upstream.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    - name: Download and Decompress the MPC Observatory Code File
+      run: |
+        python mpc_obscodes/fetch.py --compressed_file obscodes_extended.json.gz --decompressed_file obscodes_extended.json --md5_file obscodes_extended_new.md5
+    - name: Detect Changes to obscodes_extended.json
+      id: compare_step
+      run: | 
+        cp mpc_obscodes/obscodes_extended.md5 . 
+        md5sum --check obscodes_extended.md5 && echo 'upstream_change=false' >> $GITHUB_OUTPUT || echo 'upstream_change=true' >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (moeyensj) [reply .github/workflows/compare_upstream.yml:29]: I think there are a few things at work here:

1) There is actually a bug in the currently published version of the `mpc-obscodes` package: the .md5 file didn't get packaged correctly and is missing.

2) What is being compared is definitely a little confusing. I had to run through this a few times as I was working on  the workflows. So `md5sum --check` takes the output of a `md5sum {FILE}` calculation and checks the hash against the files listed in the output. Example: 
```
└─▪ md5sum obscodes_extended.json > obscodes_extended.md5 
└─▪ cat obscodes_extended.md5 
64c09eb0469e086716b102336087dd61  obscodes_extended.json
```
As long as the package already ships with its md5 hash as above it should all work as intended. I'll take a look at dropping this all into Python. 

When it comes to running the check:
```
└─▪ md5sum --check obscodes_extended.md5
obscodes_extended.json: OK
```
If I change the file name listed inside the output file from earlier:
```
└─▪ cat obscodes_extended.md5 
64c09eb0469e086716b102336087dd61  foo.json
```
```
└─▪ md5sum --check obscodes_extended.md5
md5sum: foo.json: No such file or directory
foo.json: FAILED open or read
md5sum: WARNING: 1 listed file could not be read
```

**Comment by moeyensj** [comment]
Bug fix for missing .md5 file with tests using a python based checksum has been added in #4. I'll rebase this branch once that one is merged.

**Comment by moeyensj** [comment]
I've made it so the checksums are calculated and compared purely in Python. Thanks for the suggestion @akoumjian! The workflow should be a lot cleaner now.

**Pull request merged**